flink-connector-gcp-pubsub-e2e-tests/src/test/java/org/apache/flink/connector/gcp/pubsub/sink/PubSubSinkV2ITTests.java

```diff
@@ -0,0 +1,114 @@
+import java.io.IOException;
+import java.util.List;
+
+/** Integration tests for {@link PubSubSinkV2} using {@link PubSubEmulatorContainer}. */
+@ExtendWith(MiniClusterExtension.class)
+@Execution(ExecutionMode.CONCURRENT)
+@Testcontainers
+class PubSubSinkV2ITTests extends TestLogger {
+
+    private static final String PROJECT_ID = "test-project";
+
+    private static final String TOPIC_ID = "test-topic";
+
+    private static final String SUBSCRIPTION_ID = "test-subscription";
+
+    private StreamExecutionEnvironment env;
+
+    @Container
+    private static final PubSubEmulatorContainer PUB_SUB_EMULATOR_CONTAINER =
+            new PubSubEmulatorContainer(
+                    DockerImageName.parse(DockerImageVersions.GOOGLE_CLOUD_PUBSUB_EMULATOR));
+
+    private PubsubHelper pubSubHelper;
+
+    @BeforeEach
+    void setUp() throws IOException {
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(1);
+        pubSubHelper = new PubsubHelper(PUB_SUB_EMULATOR_CONTAINER.getEmulatorEndpoint());
+
+        pubSubHelper.createTopic(PROJECT_ID, TOPIC_ID);
+        pubSubHelper.createSubscription(PROJECT_ID, SUBSCRIPTION_ID, PROJECT_ID, TOPIC_ID);
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        pubSubHelper.deleteSubscription(PROJECT_ID, SUBSCRIPTION_ID);
+        pubSubHelper.deleteTopic(PROJECT_ID, TOPIC_ID);
+        pubSubHelper.close();
+    }
+
+    @Test
+    void pubSubSinkV2DeliversRecords() throws Exception {
+        String[] elements = new String[] {"test1", "test2", "test3"};
+        DataStreamSource<String> stream =
+                env.fromSource(
+                        new DataGeneratorSource<>(
+                                new FromElementsGeneratorFunction<>(
+                                        BasicTypeInfo.STRING_TYPE_INFO, elements),
+                                elements.length,
+                                TypeInformation.of(String.class)),
+                        WatermarkStrategy.noWatermarks(),
+                        "DataGeneratorSource");
+
+        GcpPublisherConfig gcpPublisherConfig =
+                GcpPublisherConfig.builder()
+                        .setCredentialsProvider(EmulatorCredentialsProvider.create())
+                        .setTransportChannelProvider(
+                                new TestChannelProvider(
+                                        PUB_SUB_EMULATOR_CONTAINER.getEmulatorEndpoint()))
+                        .build();
+
+        PubSubSinkV2<String> sink =
+                PubSubSinkV2.<String>builder()
+                        .setProjectId(PROJECT_ID)
+                        .setTopicId(TOPIC_ID)
+                        .setSerializationSchema(new SimpleStringSchema())
+                        .setGcpPublisherConfig(gcpPublisherConfig)
+                        .setFailOnError(true)
+                        .build();
+        stream.sinkTo(sink);
+        int maxNumberOfMessages = 100;
```

Comment:
why can't we use amount equal to size of `elements` here?

Reply:
Yes, good point